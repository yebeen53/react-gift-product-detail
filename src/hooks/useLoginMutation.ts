import { useMutation } from '@tanstack/react-query';
import { apiClient } from '@/api/apiClient';

interface LoginPayload {
  email: string;
  password: string;
}

interface LoginResponse {
  authToken: string;
  id: number;
  name: string;
  email: string;
}

export const useLoginMutation = () => {
  return useMutation({
    mutationFn: async (payload: LoginPayload): Promise<LoginResponse> => {
      const res = await apiClient.post('/api/login', payload);
      return res.data.data;
    },
  });
};