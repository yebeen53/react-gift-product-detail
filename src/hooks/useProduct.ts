import {
  useSuspenseQuery,
  useMutation,
  useQueryClient,
} from '@tanstack/react-query';
import {
  fetchProduct,
  fetchProductDetail,
  fetchHighlightReview,
  fetchWishCount,
} from '@/api/products/product';

export const useProduct = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['product', productId],
    queryFn: () => fetchProduct(productId),
  });

export const useProductDetail = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['productDetail', productId],
    queryFn: () => fetchProductDetail(productId),
  });

export const useHighlightReview = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['highlightReview', productId],
    queryFn: () => fetchHighlightReview(productId),
  });

export const useWishCount = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['wishCount', productId],
    queryFn: () => fetchWishCount(productId),
  });

export const useWishMutation = (productId: string) => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async () => Promise.resolve(), // 실제 API 요청 함수 넣기

    onMutate: async () => {
      await queryClient.cancelQueries({ queryKey: ['wishCount', productId] });

      const prevWish = queryClient.getQueryData<{
        wishCount: number;
        isWished: boolean;
      }>(['wishCount', productId]);

      if (prevWish) {
        queryClient.setQueryData(['wishCount', productId], {
          ...prevWish,

          wishCount: prevWish.isWished
            ? prevWish.wishCount - 1
            : prevWish.wishCount + 1,
          isWished: !prevWish.isWished, // 토글
        });
      }

      return { prevWish };
    },

    onError: (_err, _vars, context) => {
      if (context?.prevWish) {
        queryClient.setQueryData(['wishCount', productId], context.prevWish);
      }
    },

    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['wishCount', productId] });
    },
  });
};
