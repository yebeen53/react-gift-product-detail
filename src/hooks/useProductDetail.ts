import { useSuspenseQuery } from '@tanstack/react-query';
import axios from 'axios';

interface Announcement {
  name: string;
  value: string;
  displayOrder: number;
}

interface ProductDetail {
  description: string;
  announcements: Announcement[];
}

const fetchProductDetail = async (productId: string): Promise<ProductDetail> => {
  const response = await axios.get(`/api/products/${productId}/detail`);
  return response.data.data;
};

export const useProductDetail = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['productDetail', productId],
    queryFn: () => fetchProductDetail(productId),
  });
