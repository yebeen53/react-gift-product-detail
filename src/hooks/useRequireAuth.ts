import useAuth from '@/hooks/useAuth';
import { useLocation, useNavigate } from 'react-router-dom';
import { useEffect } from 'react';
import { ROUTES } from '@/constants/routes';

const useRequireAuth = () => {
  const { user } = useAuth();
  const location = useLocation();
  const navigate = useNavigate();

  useEffect(() => {
    if (!user || !user.authToken) {
      navigate(ROUTES.HOME, {
        replace: true,
        state: { from: location.pathname },
      });
    }
  }, [user, navigate, location]);

  return user;
};

export default useRequireAuth;
