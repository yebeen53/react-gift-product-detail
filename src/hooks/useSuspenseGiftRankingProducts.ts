import { fetchRanking } from '@/api/products/ranking';
import { useSuspenseQuery } from '@tanstack/react-query';

export const useSuspenseGiftRankingProducts = (
  genderCode: string,
  categoryCode: string
) => {
  return useSuspenseQuery({
    queryKey: ['giftRankingProducts', genderCode, categoryCode],
    queryFn: () => fetchRanking(genderCode, categoryCode),
  });
};
