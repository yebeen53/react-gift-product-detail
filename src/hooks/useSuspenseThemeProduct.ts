import {
  useSuspenseInfiniteQuery,
  type QueryFunctionContext,
} from '@tanstack/react-query';
import axios from 'axios';
import type { ThemeProductsData } from '@/types/product';

const fetchThemeProducts = async (
  ctx: QueryFunctionContext<readonly [string, string], number>
): Promise<ThemeProductsData> => {
  const [, themeId] = ctx.queryKey;
  const pageParam = ctx.pageParam ?? 0;

  const res = await axios.get(`/api/themes/${themeId}/products`, {
    params: { cursor: pageParam, limit: 10 },
  });
  const data = res.data.data;

  return {
    list: data.list,
    cursor: data.cursor,
    hasMoreList: data.hasMoreList,
  };
};

export const useSuspenseThemeProducts = (themeId: string) => {
  return useSuspenseInfiniteQuery({
    queryKey: ['themeProducts', themeId] as const,
    queryFn: fetchThemeProducts,
    initialPageParam: 0,
    getNextPageParam: (lastPage) =>
      lastPage.hasMoreList ? lastPage.cursor : undefined,
  });
};
