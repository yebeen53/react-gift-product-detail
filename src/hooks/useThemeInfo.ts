import { apiClient } from '@/api/apiClient';
import { useSuspenseQuery } from '@tanstack/react-query';

export interface ThemeDetail {
  image: string | undefined;
  name: string;
  themeId: string;
  title: string;
  description: string | null;
  backgroundColor: string;
}

const fetchThemeDetail = async (themeId: string): Promise<ThemeDetail> => {
  const res = await apiClient.get(`/api/themes/${themeId}/info`);
  const data = res.data?.data;
  if (!data) {
    throw new Error('테마 정보가 없습니다.');
  }
  return data;
};

export const useThemeInfo = (themeId: string) => {
  return useSuspenseQuery({
    queryKey: ['themeDetail', themeId],
    queryFn: () => fetchThemeDetail(themeId),
  });
};
