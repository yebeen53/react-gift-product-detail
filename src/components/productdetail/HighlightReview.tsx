import theme from '@/data/theme';
import { useHighlightReview } from '@/hooks/useProduct';

interface Review {
  id: string;
  authorName: string;
  content: string;
}

interface Props {
  productId: string;
}

const HighlightReview = ({ productId }: Props) => {
  const { data: reviewData } = useHighlightReview(productId);

  if (!reviewData) return null;

  return (
    <section style={{ margin: theme.spacing.spacing4 }}>
      <h2>선물후기</h2>
      {reviewData.reviews?.map((review: Review) => (
        <div key={review.id}>
          <p>{review.content}</p>
          <p>{review.authorName}</p>
        </div>
      ))}
    </section>
  );
};

export default HighlightReview;
