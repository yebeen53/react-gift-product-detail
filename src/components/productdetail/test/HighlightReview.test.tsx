import { render, screen} from '@testing-library/react';
import HighlightReview from '../HighlightReview';
import { useHighlightReview } from '@/hooks/useProduct'; 


jest.mock('@/hooks/useProduct', () => ({
  useHighlightReview: jest.fn(),
}));

describe('HighlightReview 컴포넌트', () => {
  const mockReviewData = {
    reviews: [
      { id: '1', authorName: '김구매자', content: '정말 만족스러운 상품입니다!' },
      { id: '2', authorName: '이소비자', content: '배송이 빠르고 품질도 좋아요.' },
    ],
  };

  beforeEach(() => {
    (useHighlightReview as jest.Mock).mockClear();
  });

  it('리뷰 데이터가 없을 때 아무것도 렌더링하지 않아야 한다', () => {
    (useHighlightReview as jest.Mock).mockReturnValue({ data: undefined });
    const { container } = render(<HighlightReview productId="p1" />);
    expect(container).toBeEmptyDOMElement();
  });

  it('리뷰 데이터가 있을 때 제목과 리뷰 내용이 올바르게 렌더링되어야 한다', async () => {
    (useHighlightReview as jest.Mock).mockReturnValue({ data: mockReviewData });
    render(<HighlightReview productId="p1" />);

    expect(screen.getByRole('heading', { name: '선물후기' })).toBeInTheDocument();
    expect(screen.getByText('정말 만족스러운 상품입니다!')).toBeInTheDocument();
    expect(screen.getByText('김구매자')).toBeInTheDocument();
    expect(screen.getByText('배송이 빠르고 품질도 좋아요.')).toBeInTheDocument();
    expect(screen.getByText('이소비자')).toBeInTheDocument();
  });

  it('리뷰 배열이 비어있을 때 제목만 렌더링되어야 한다', () => {
    (useHighlightReview as jest.Mock).mockReturnValue({ data: { reviews: [] } });
    render(<HighlightReview productId="p1" />);
    expect(screen.getByRole('heading', { name: '선물후기' })).toBeInTheDocument();
    expect(screen.queryByText('정말 만족스러운 상품입니다!')).not.toBeInTheDocument();
  });

  it('컴포넌트의 스타일 및 구조가 스냅샷과 일치해야 한다', () => {
    (useHighlightReview as jest.Mock).mockReturnValue({ data: mockReviewData });
    const { asFragment } = render(<HighlightReview productId="p1" />);
    expect(asFragment()).toMatchSnapshot();
  });
});