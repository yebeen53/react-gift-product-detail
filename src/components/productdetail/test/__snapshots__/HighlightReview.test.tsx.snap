// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`HighlightReview 컴포넌트 컴포넌트의 스타일 및 구조가 스냅샷과 일치해야 한다 1`] = `
<DocumentFragment>
  <section
    style="margin: 16px;"
  >
    <h2>
      선물후기
    </h2>
    <div>
      <p>
        정말 만족스러운 상품입니다!
      </p>
      <p>
        김구매자
      </p>
    </div>
    <div>
      <p>
        배송이 빠르고 품질도 좋아요.
      </p>
      <p>
        이소비자
      </p>
    </div>
  </section>
</DocumentFragment>
`;
